src/common/linearising_executor.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Christopher James Halse Rogers <christopher.halse.rogers@canonical.com>
+ */
+
+#include "mir/linearising_executor.h"
+#include "mir/system_executor.h"
+
+#include <deque>
+#include <mutex>
+#include <thread>
+
+namespace
+{
+class LinearisingAdaptor : public mir::Executor
+{
+public:
+    LinearisingAdaptor() noexcept
+        : idle{true}
+    {
+    }
+
+    ~LinearisingAdaptor() noexcept
+    {
+        std::unique_lock<decltype(mutex)> lock{mutex};
+        workqueue.clear();
+        while (!idle)
+        {
+            lock.unlock();
+            std::this_thread::sleep_for(std::chrono::milliseconds{1});
+            lock.lock();
+        }
```

Comment:
This looks like a job for `std::condition_variable::wait_for` is there an advantage to spinning?